homeassistant/components/reolink/media_source.py

```diff
@@ -0,0 +1,329 @@
+"""Expose Reolink IP camera VODs as media sources."""
+
+from __future__ import annotations
+
+import datetime as dt
+import logging
+
+from homeassistant.components.camera import DOMAIN as CAM_DOMAIN, DynamicStreamSettings
+from homeassistant.components.media_player import MediaClass, MediaType
+from homeassistant.components.media_source.error import Unresolvable
+from homeassistant.components.media_source.models import (
+    BrowseMediaSource,
+    MediaSource,
+    MediaSourceItem,
+    PlayMedia,
+)
+from homeassistant.components.stream import create_stream
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_get_media_source(hass: HomeAssistant) -> ReolinkVODMediaSource:
+    """Set up camera media source."""
+    return ReolinkVODMediaSource(hass)
+
+
+def res_name(stream: str) -> str:
+    """Return the user friendly name for a stream."""
+    return "High res." if stream == "main" else "Low res."
+
+
+class ReolinkVODMediaSource(MediaSource):
+    """Provide Reolink camera VODs as media sources."""
+
+    name: str = "Reolink"
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize ReolinkVODMediaSource."""
+        super().__init__(DOMAIN)
+        self.hass = hass
+        self.data = hass.data[DOMAIN]
```

Comment:
Would be nice to type `self.data` as well since its nearly the only thing not typed here

Reply:
`dict[str, ????]`

Reply:
done, its `dict[str, ReolinkData]`